### PR TITLE
Fix warm caching failing due to a wrong assumption around Mac Catalyst

### PR DESCRIPTION
**Pull request by pepicrft**
### Short description 📝
While testing the cache, @Alex-Ozun noted that Tuist yielded "X not found" with a Swift Package integrated through `Dependencies.swift`. After some debugging, it turns out that our warming logic assumed that the only condition to build for  Mac Catalyst is **having some** frameworks that support Catalyst. This is assumption has proven invalid. In his case, which can easily be reproduced [here](https://github.com/Alex-Ozun/TuistCacheRepro), the package dependency supports Mac Catalyst but the dependent target not, and that caused the issue.

To fix it, I extended the logic to traverse the graph when determining if a target can be built for Mac Catalyst. In the future, we might want to iterate on the logic to make it more granular, but that'll come at the cost of some complexity in the logic that generates the schemes because we no longer can have a large scheme that aggregates all the binary targets together.

**Note** that the logic to determine if a dependency supports Mac Catalyst only works with source targets. In another PR I'll extend the logic to obtain that information from pre-compiled binaries such as libraries, frameworks, and xcframeworks.

### How to test the changes locally 🧐
You should be able to successfully run `swift run tuist cache warm -x` against [this project](https://github.com/Alex-Ozun/TuistCacheRepro) and then generate a project passing the `-x` flag and get the dependency as a binary and a project that compiles.

### Contributor checklist ✅

- [x] The code has been linted using run `./fourier lint tuist --fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/TuistCore/Graph/GraphTraverser.swift [hunk 1]
```diff
@@ -181,6 +181,19 @@ public class GraphTraverser: GraphTraversing {
             .first { $0.target.product == .appClip }
     }
 
+    public func buildsForMacCatalyst(path: AbsolutePath, name: String) -> Bool {
+        let dependenciesSupportMacCatalyst = allDependenciesSatisfy(from: .target(name: name, path: path)) { dependency in
+            if let target = self.target(from: dependency) {
+                return target.target.supportsCatalyst
+            } else {
+                // TODO: - Obtain this information from pre-compiled binaries
```
Comment (waltflanagan) [inline Sources/TuistCore/Graph/GraphTraverser.swift:189]: We'll likely need this logic for other area related to multi-platform dependencies. Not sure what it should look like or where it should live, but wanted to call it out.

**Review by waltflanagan — APPROVED** [review]

File: Sources/TuistCore/Graph/GraphTraverser.swift [hunk 2]
```diff
@@ -565,6 +578,19 @@ public class GraphTraverser: GraphTraversing {
...
+        var traversedDependenciesThatMeetCriteria: UInt = 0
+        _ = filterDependencies(from: rootDependency, test: { dependency in
+            traversedDependencies += 1
+            if meets(dependency) {
+                traversedDependenciesThatMeetCriteria += 1
+            }
+            return true
+        })
+        return traversedDependencies == traversedDependenciesThatMeetCriteria
+    }
```
Comment (Alex-Ozun) [inline Sources/TuistCore/Graph/GraphTraverser.swift:592]: nitpick
```suggestion
    func allDependenciesSatisfy(from rootDependency: GraphDependency, meets: (GraphDependency) -> Bool) -> Bool {
        var allSatisfy = true
        _ = filterDependencies(from: rootDependency, test: { dependency in
            if !meets(dependency) {
               allSatisfy = false
            }
            return true
        })
        return allSatisfy
```

**Review by Alex-Ozun — APPROVED** [review]
Thanks @pepicrft !

File: Sources/TuistCore/Graph/GraphTraverser.swift [hunk 3]
```diff
@@ -181,6 +181,19 @@ public class GraphTraverser: GraphTraversing {
...
+        let dependenciesSupportMacCatalyst = allDependenciesSatisfy(from: .target(name: name, path: path)) { dependency in
+            if let target = self.target(from: dependency) {
+                return target.target.supportsCatalyst
+            } else {
+                // TODO: - Obtain this information from pre-compiled binaries
+                // lipo -info should include "macabi" in the list of architectures
+                return false
+            }
+        }
+        return dependenciesSupportMacCatalyst && (target(path: path, name: name)?.target.supportsCatalyst ?? false)
```
Comment (danyf90) [inline Sources/TuistCore/Graph/GraphTraverser.swift:194]: nit: maybe we can check `target(path: path, name: name)?.target.supportsCatalyst` at the beginning and return earlier without processing the whole three if false?

**Review by danyf90 — APPROVED** [review]

File: Sources/TuistCore/Graph/GraphTraverser.swift (hunk 1 above)
Comment (pepicrft) [reply Sources/TuistCore/Graph/GraphTraverser.swift:189]: I'll ping you once it's ready.

File: Sources/TuistCore/Graph/GraphTraverser.swift (hunk 3 above)
Comment (pepicrft) [reply Sources/TuistCore/Graph/GraphTraverser.swift:194]: Very good point. There's no need to traverse the graph if the target we are traversing from doesn't support it. I'll push a fix.

**Review by danyf90 — APPROVED** [review]

**Pull request merged**